### ISSUE TITLE
deepEqual compare nextProps and old props do not work properly

### ISSUE THREAD

**Issue by bintoll** [bkrem/react-d3-tree#59]
Tried:
Pass data to Tree component as state.
Render the tree.
Add children to tree data and write in state.
Nothing happens with Tree.

I found why it does not work:
it is used deepEqual(this.props.data, nextProps.data) (i also tried a strict mode)
if i change it to this.props.data !== nextProps.data - the tree component updates correctly.

- [ ] Implementation
- [ ] Unit test
- [ ] Docs

**Comment by bkrem** [comment]
Hey, sorry it's taken me so long to respond, I've not been able to commit to any substantial open source work for the past few weeks.

Could you please create a quick reproduction case on https://codesandbox.io/ or similar?

**Comment by dclemenzi** [comment]
Ran into the same issue and came to the same conclusion that the deepEqual is the cause. My solution perhaps less elegant was to clone the data which causes the tree to update correctly. 

var data = clone(this.state.data)
...
Do Work
...
this.setState({data: data})

But this is now causing issues with thrashing the layout on large amount of updates. Is there a correct way to use this Tree as a dynamic view feeding it data?

**Comment by bkrem** [comment]
Hi @dclemenzi,

Sorry to hear you experienced a similar issue.

Could you give me a quick example of what the diff between the `this.state.data` and `nextState.data` would be in your case, or reproduce a simplified version here https://codesandbox.io/s/ykn9o96nm9 ?

**Comment by dclemenzi** [comment]
Ok I have recreated the issue I am having in the sandbox. There is a react-interval running that adds a new child every second and its working because I am cloning this.state.data on each insert. If you comment out line 39 and add back in line 40 which removes the clone it stops working. 

I am probably doing something wrong and would love to know the correct way to dynamically insert nodes without having to do a deep clone. Any insight would be greatly appreciated. Thanks!!

https://codesandbox.io/s/2vkxr7rryn

**Comment by bkrem** [comment]
@dclemenzi thank you for taking the time to create a repro case! 

Yeah this is definitely unintended behaviour. You're not doing anything wrong, I simply failed to test this use case sufficiently.

Scheduling this as the highest prio change since it's literally just the `deepEqual` check having unintended consequences.
Will try to release a fixed version by this evening or tomorrow.

**Comment by dclemenzi** [comment]
Looking forward to trying you fix!

**Comment by bkrem** [comment]
Fix shipped in [v1.10.1](https://github.com/bkrem/react-d3-tree/releases/tag/v1.10.1)  🚀 

@dclemenzi you should now simply be able to do:
```js
const data = [...this.state.data]
// ... as before
```

but I would still recommend using `clone` for your use case.

I'm going to do a quick post-mortem to explain why for future reference, since I'm sure you two won't be the last who come across this:

The issue here was partially at the library level (i.e. react-d3-tree) and partially at the language level (JavaScript).
In your repro the broken case was roughly:
```js
const data = this.state.data // `data` now simply points to `this.state.data`
data.children[0].push(...) // `this.state.data` is mutated
this.setState({data})
```
This meant that since `this.state.data` is passed to `react-d3-tree` as the `data` prop, the `deepEqual` call in the tree's `componentWillReceiveProps` was actually correct.
Since `this.props.data` was mutated and then **passed a reference to itself** as `nextProps.data`, this meant that `this.props.data === nextProps.data`, both by value and reference.

Replacing the `deepEqual` check with an identity check (`===`) now allows the tree's data to be updated simply if the top-level array is a different one (i.e. `const data = [...this.state.data]`). The underlying issue remains the same though, since **all the nested objects still point to `this.state.data` and thus the tree's `this.props.data`**.

While this currently works fine, I'd still highly recommend taking an immutable approach to the data you pass to the tree (i.e. `clone` the whole `data` array to unlink JS pointer references -> mutate the dataset -> update your state that you pass to the tree).

For more context on call-by-value vs. call-by-reference in JS check [this](https://medium.com/@naveenkarippai/learning-how-references-work-in-javascript-a066a4e15600).

Hope that helps!
